icefall/checkpoint.py

```diff
@@ -327,3 +343,95 @@ def remove_checkpoints(
     for c in to_remove:
         os.remove(c)
+
+
+def update_averaged_model(
+    params: Dict[str, Tensor],
+    model_cur: Union[nn.Module, DDP],
+    model_avg: nn.Module,
+) -> None:
+    """Update the averaged model,
+
+    Args:
+      params:
+        User defined parameters, e.g., epoch, loss.
+      model_cur:
+        The current model.
+      model_avg:
+        The stored model averaged from start of training to update.
+    """
+    weight_cur = params.average_period / params.batch_idx_train
+    weight_avg = 1 - weight_cur
+
+    if isinstance(model_cur, DDP):
+        model_cur = model_cur.module
+
+    cur = model_cur.state_dict()
+    avg = model_avg.state_dict()
+
+    uniqued: Dict[int, str] = dict()
+    for k, v in avg.items():
+        v_data_ptr = v.data_ptr()
+        if v_data_ptr in uniqued:
+            continue
+        uniqued[v_data_ptr] = k
+
+    uniqued_names = list(uniqued.values())
+    for k in uniqued_names:
+        avg[k] *= weight_avg
+        avg[k] += cur[k] * weight_cur
+
+
+def average_checkpoints_with_averaged_model(
+    filename_start: str,
+    filename_end: str,
+    device: torch.device = torch.device("cpu"),
+) -> Dict[str, Tensor]:
+    """Average model parameters over the range with given
+    start model(excluded) and end model.
+
+    Let start = batch_idx_train of model-start,
+        end = batch_idx_train of model-end,
+    Then the average model over epoch [start+1, start+2, ..., end] is
+        avg = (model_end * end - model_start * start) / (start - end)
+
+    The model index could be epoch number or checkpoint number.
+
+    Args:
+      filename_start:
+        Checkpoint filename of the start model. We assume it
+        is saved by :func:`save_checkpoint`.
+      filename_end:
+        Checkpoint filename of the end model. We assume it
+        is saved by :func:`save_checkpoint`.
+      device:
+        Move checkpoints to this device before averaging.
+    """
+    state_dict_start = torch.load(filename_start, map_location=device)
+    state_dict_end = torch.load(filename_end, map_location=device)
+
+    batch_idx_train_start = state_dict_start["batch_idx_train"]
+    batch_idx_train_end = state_dict_end["batch_idx_train"]
+    interval = batch_idx_train_end - batch_idx_train_start
+    weight_start = -batch_idx_train_start / interval
+    weight_end = batch_idx_train_end / interval
+
+    model_end = state_dict_end["model_avg"]
+    model_start = state_dict_start["model_avg"]
+    avg = model_end
+
+    # Identify shared parameters. Two parameters are said to be shared
+    # if they have the same data_ptr
+    uniqued: Dict[int, str] = dict()
+    for k, v in avg.items():
+        v_data_ptr = v.data_ptr()
+        if v_data_ptr in uniqued:
+            continue
+        uniqued[v_data_ptr] = k
+
+    uniqued_names = list(uniqued.values())
+    for k in uniqued_names:
+        avg[k] *= weight_end
+        avg[k] += model_start[k] * weight_start
+
```

Comment:
This part is almost the same as the above function. Please refactor it to reduce redundant code.